intel-sgx/async-usercalls/src/queues.rs

```diff
@@ -0,0 +1,209 @@
+            if let Some(sender) = provider_map.get(ret.provider_id()).and_then(|entry| entry.as_ref()) {
+                let _ = sender.send(*ret);
+            }
+        }
+    }
+
+    fn run(self) {
+        let mut returns = [Identified::default(); Self::RECV_BATCH_SIZE];
+        loop {
+            // Block until there is a return. Then we receive any other values
+            // from the return queue **without** blocking using `try_iter()`.
+            let first = match self.return_queue_rx.recv() {
+                Ok(ret) => ret,
+                Err(RecvError::Closed) => break,
+            };
+            let mut count = 0;
+            for ret in iter::once(first).chain(self.return_queue_rx.try_iter().take(Self::RECV_BATCH_SIZE - 1)) {
+                assert!(ret.id != 0);
+                returns[count] = ret;
+                count += 1;
+            }
+            self.send(&returns[..count]);
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+enum Queue {
+    Usercall,
+    Return,
+    Cancel,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct QueueSynchronizer {
+    queue: Queue,
+}
+
+impl Synchronizer for QueueSynchronizer {
+    fn wait(&self, event: QueueEvent) -> Result<(), SynchronizationError> {
+        let ev = match (self.queue, event) {
+            (Queue::Usercall, QueueEvent::NotEmpty) => panic!("enclave should not recv on usercall queue"),
+            (Queue::Cancel, QueueEvent::NotEmpty) => panic!("enclave should not recv on cancel queue"),
+            (Queue::Return, QueueEvent::NotFull) => panic!("enclave should not send on return queue"),
+            (Queue::Usercall, QueueEvent::NotFull) => EV_USERCALLQ_NOT_FULL,
+            (Queue::Cancel, QueueEvent::NotFull) => EV_CANCELQ_NOT_FULL,
+            (Queue::Return, QueueEvent::NotEmpty) => EV_RETURNQ_NOT_EMPTY,
+        };
+        unsafe {
+            raw::wait(ev, raw::WAIT_INDEFINITE);
+        }
+        Ok(())
+    }
+
+    fn notify(&self, _event: QueueEvent) {
+        // any synchronous usercall would do
+        unsafe {
+            raw::wait(0, raw::WAIT_NO);
+        }
+    }
+}
+
+mod map {
+    use fnv::FnvHashMap;
+
+    pub struct SequentialMap<T> {
+        map: FnvHashMap<u32, T>,
+        next_id: u32,
+    }
+
+    impl<T> SequentialMap<T> {
+        pub fn new() -> Self {
+            Self {
+                map: FnvHashMap::with_capacity_and_hasher(16, Default::default()),
+                next_id: 0,
+            }
+        }
+
+        /// Insert a new value and return the index.
+        ///
+        /// # Panics
+        /// Panics if `u32::MAX` entries are kept in the queue and a new one is attempted
+        /// to be inserted. Note that removing entries from the queue also frees up space
+        /// and the number of available entries.
+        pub fn insert(&mut self, value: T) -> u32 {
+            let initial_id = self.next_id;
+            loop {
+                let id = self.next_id;
```

Comment:
There's an extremely unlikely but severe potential performance problem with this approach for allocating ids. Suppose you have `2^32 - 1` calls that never terminate, and then a single other call which is fast but you never have more than one of those outstanding at once. Each time you make the fast call, you'll need to do 2^32 map lookups to find the available id.

I don't think we need to worry about this in practice.

Reply:
But it does make me wonder if we might want to have some way to limit the length of the call queue to something smaller than 2^32, with either blocking if the queue is full, or switching to making the call synchronously in this case. This might not be useful for DSM, but I can imagine some type of EDP application in which enclave threads could be generating asynchronous calls faster than they could be serviced, and you might want to have the ability to either throttle the enclave threads when the queue gets full or switch to synchronous ocalls, instead of growing the queue without bounds and getting a panic when you run out of memory.

Reply:
I agree this is a potential problem, but can/should be fixed as a separate issue. @arai-fortanix what do you think?

Reply:
Fixing later seems fine.

Reply:
Created issue #550 for this